Support grpc-retry-pushback-ms (#664)
open-telemetry/opentelemetry-proto: docs/specification.md

and MAY supply additional
[details via status](https://godoc.org/google.golang.org/grpc/status#Status.WithDetails)
using
[RetryInfo](https://github.com/googleapis/googleapis/blob/6a8c7914d1b79bd832b5157a09a9332e8cbd16d4/google/rpc/error_details.proto#L40)

> or via Trailer [with the gRPC metadata key `grpc-retry-pushback-ms`](https://github.com/grpc/proposal/blob/master/A6-client-retries.md#pushback).

Reviewer: Please link to a permanent link instead of one that can break.

Reviewer 2: Isn’t the problem that permanent link may include outdated information?

The client SHOULD honor at least one of these backpressure mechanisms.

> In order to have the backpressure respected, the server SHOULD implement both mechanisms.

Reviewer: Does this mean the server should be responding with both mechanisms at the same time? The example below only has one mechanism being used at a time.

This seems odd to me. It is usually the client that is recommended to handle multiple versions of a response and the server is expected to support at least one.

Is there a preference for which mechanism the server should support?

Author: I don't think there is a preference. About half the languages are using gRPC retry config which supports only the `pushback-ms` header, and the other half is rolling their own retry logic and only support the `RetryInfo`..

If a client uses gRPC retry config, it is offloading all the retry logic to gRPC internal library, so it can't also support `RetryInfo`.

    }
  }
```
Here is a snippet of sample Python code to illustrate using `grpc-retry-pushback-ms`:

```python
  # Do this on the server side.
  context.set_trailing_metadata(
        (
            (
                "grpc-retry-pushback-ms",
                "5000", # 5 seconds
            ),
        )
    )
  context.abort(StatusCode.UNAVAILABLE, "")

  ...

  # On the client side use gRPC retry config: https://grpc.io/docs/guides/retry/, which will automatically parse this header
  # and handle all backoff and retry logic.
  # Or manually parse the header:
  
  pushback = dict(error.trailing_metadata()).get("grpc-retry-pushback-ms")
  if pushback is not None:

>     wait_period = int(pushback)

Reviewer 3: I think that the gRPC clients are supposed to handle this out-of-the-box. From https://github.com/grpc/proposal/blob/master/A6-client-retries.md#pushback

> When a client receives an explicit pushback response from a server, and it is appropriate to retry the RPC, it will retry after exactly that delay.

Author: >> I think if they do NOT use gRPC retry config then they should not parse grpc-retry-pushback-ms as it is part of the gRPC retry mechanism.

Why not though ?

Reviewer 3: If the gRPC client in given language (e.g. new version) would start doing it then you could do the retry (exponentially?) more times: first by gRPC client and then second time by custom `grpc-retry-pushback-ms ` handling in the exporter implementation. It is safer to leave the `grpc-retry-pushback-ms` implementation to the gRPC clients.

```

When the client receives this signal, it SHOULD follow the recommendations

> outline in the documentation for [`grpc-retry-pushback-ms`](https://github.com/grpc/proposal/blob/master/A6-client-retries.md#pushback)

Reviewer 3: This is already a specification (design) of gRPC clients.

    }
  }
```
Here is a snippet of sample Python code to illustrate using `grpc-retry-pushback-ms`:

```go
# Do this on the server side.
trailer := metadata.Pairs("grpc-retry-pushback-ms", "5000")
grpc.SetTrailer(ctx, trailer)
```

On the client side using gRPC retry config (https://grpc.io/docs/guides/retry/) 

> will cause gRPC to automatically parse this header and handle all backoff and retry logic.

Reviewer 3: ```suggestion
will cause the gRPC client to automatically parse the `grpc-retry-pushback-ms` trailer metadata
and handle all backoff and retry logic.
```

> When the client receives this signal, if it isn't using gRPC retry config, it SHOULD 
> follow the recommendations or outlined in documentation for
> [RetryInfo](https://github.com/googleapis/googleapis/blob/6a8c7914d1b79bd832b5157a09a9332e8cbd16d4/google/rpc/error_details.proto#L40).

Reviewer 3: > if it isn't using gRPC retry config

This is not clear for me.
Does it suggest that the client should support only one retry pushback mechanism?
Would it be bad to support both of them?

How about

```suggestion
The description of how the client should respect `RetryInfo` is provided
in its [documentation](https://github.com/googleapis/googleapis/blob/6a8c7914d1b79bd832b5157a09a9332e8cbd16d4/google/rpc/error_details.proto#L34-L39).
```

    }
  }
```

> Here is a snippet of sample Go code to illustrate using `grpc-retry-pushback-ms`:

Reviewer 3: nit: Add blank line. Otherwise some linters will complain

```suggestion

Here is a snippet of sample Go code to illustrate using `grpc-retry-pushback-ms`:
```

  }
```

Here is a snippet of sample Go code to illustrate using `grpc-retry-pushback-ms`:

```go
// Do this on the server side.

> trailer := metadata.Pairs("grpc-retry-pushback-ms", "5000")

Reviewer 4: Is this expected to be done _only_ if the server wants to pushback? It is not clear from the sample if this is some sort of permanent option that the server must always set or only set when the server is overloaded and needs to pushback the clients.
